ballerina/in_memory_datasource.bal

```diff
@@ -0,0 +1,162 @@
+|};
+
+public isolated client class InMemoryDatasource {
+    *datasource:Datasource;
+
+    private final table<datasource:Supergraph> key(version) supergraphTable;
+    private final table<datasource:Subgraph> key(version, name) subgraphTable;
+    private final table<SupergraphSubgraph> key(supergraphVersion, subgraphName) supergraphSubgraphTable;
+
+    public function init() {
+        self.supergraphTable = table [];
+        self.subgraphTable = table [];
+        self.supergraphSubgraphTable = table [];
+    }
+
+    isolated resource function get supergraphs() returns datasource:Supergraph[]|datasource:Error {
+        lock {
+            return self.supergraphTable.toArray().clone();
+        }
+    }
+
+    isolated resource function get supergraphs/[string version]() returns datasource:Supergraph|datasource:Error {
+        lock {
+            if !self.supergraphTable.hasKey(version) {
+                return error datasource:Error(string `No supergraph found with version '${version}'.`);
+            }
+            return self.supergraphTable.get(version).clone();
+        }
+    }
+
+    isolated resource function get supergraphs/[string version]/subgraphs() returns datasource:Subgraph[]|datasource:Error {
+        lock {
+            table<datasource:Subgraph> tableResult = from var 'join in self.supergraphSubgraphTable
+                                                     from var subgraph in self.subgraphTable
+                                                     where 'join.supergraphVersion === version && 'join.subgraphVersion === subgraph.version && 'join.subgraphName === subgraph.name
+                                                     select subgraph;
+            return tableResult.clone().toArray();
+        }
+    }
+
+    isolated resource function post supergraphs(datasource:SupergraphInsert data) returns datasource:Error? {
+        lock {
+            if self.supergraphTable.hasKey(data.version) {
+                return error datasource:Error(string `A supergraph already exists with the given version '${data.version}'`);
+            }
+            datasource:Supergraph supergraph = {
+                version: data.version,
+                schema: data.schema,
+                apiSchema: data.apiSchema
+            };
+            self.supergraphTable.add(supergraph);
+            foreach datasource:SubgraphId id in data.clone().subgraphs {
+                self.supergraphSubgraphTable.add({ 
+                    supergraphVersion: data.version,
+                    subgraphVersion: id.version,
+                    subgraphName: id.name 
+                });
+            }
+        }
+    }
+
+    isolated resource function put supergraphs/[string version](datasource:SupergraphUpdate data) returns datasource:Error? {
+        lock {
+            datasource:Supergraph updatedSupergraph = {
+                version,
+                schema: data.schema,
+                apiSchema: data.apiSchema
```

Comment:
The `data` (of type `datasource:SupergraphUpdate`) has some additional fields in it that are not defined in the `datasource:Supergraph`. And I have defined the `datasource:Supergraph` type as a closed record. Because almost always, it is used as an output type and not as an input by the user.
```ballerina
public type Supergraph record {|
    readonly string version;
    string schema;
    string apiSchema;
|};

public type SupergraphUpdate record {
    *Supergraph;
    SubgraphId[] subgraphs;
};
```
WDYT? Shall I change it?

Reply:
Let's keep it as it is.